### PR TITLE
Add torch.cond support to AOT Inductor

### PR DESCRIPTION
**Pull request by aakhundov**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121120

Summary: In this PR, `torch.cond` support and the necessary codegening infrastructure is added to C++ wrapper (AOTInductor and friends).

Notable additions:

- A new mechanism in the Python wrapper codegen to precompile and save the Triton kernels (generated and user-defined) which haven't been covered by the active path through the control flow given the sample inputs. As we can't do the runtime autotuning of the kernels outside the active path, we precompile and save them with the `launchers[0]` (corresponding to the first config).

- Codegen infra for `torch.cond` in the C++ wrapper (ABI- and non-ABI-compatible). The `torch.cond` codegen has been slightly refactored to avoid duplication across the Python and C++ wrappers.

- More extensions of the caching sites in the wrapper code to cache per codegened graph (e.g., `codegen_int_array_var`) + some infra for tracking the current codegened graph in the wrapper (both during codegen-ing in the `Scheduler.codegen` and in the `WrapperCodeGen.generate` functions).

- New unit tests to cover the added AOT Inductor + `torch.cond` functionality.

Codegen examples from the new unit tests:

- [`test_cond_simple_abi_compatible_cpu`](https://gist.github.com/aakhundov/862d5de9aa460f5df399e1387f7b342e)
- [`test_cond_simple_abi_compatible_cuda`](https://gist.github.com/aakhundov/d70b81f95fa8cc768cedef9acacb25bb)
- [`test_cond_simple_non_abi_compatible_cpu`](https://gist.github.com/aakhundov/c0ae7a8cbb6fa311c838e1b580f9a3f6)
- [`test_cond_simple_non_abi_compatible_cuda`](https://gist.github.com/aakhundov/08b945d4e8a32c97b7f9ff6272f4a223)
- [`test_cond_nested_abi_compatible_cuda`](https://gist.github.com/aakhundov/ce664f433c53e010ce4c0d96a6c13711)
- [`test_cond_with_parameters_abi_compatible_cuda`](https://gist.github.com/aakhundov/77afbeb8eaab5c5b930a3f922a7baf12)
- [`test_cond_with_multiple_outputs_abi_compatible_cuda`](https://gist.github.com/aakhundov/8cc06105ec8a3fe88be09b3f6e32c690)

Test Plan:

```
$ python test/inductor/test_aot_inductor.py -k test_cond
...
----------------------------------------------------------------------
Ran 42 tests in 170.619s

OK
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/cpp_wrapper_cpu.py
Comment (zou3519) [inline torch/_inductor/codegen/cpp_wrapper_cpu.py:1475]: Do you have an example of what the output code looks like?

Comment (aakhundov) [reply torch/_inductor/codegen/cpp_wrapper_cpu.py:1475]: Sure, for this model:

```python
        class SimpleCondModel(torch.nn.Module):
            def forward(self, p, a, b):
                def true_fn(x, y):
                    return x + y

                def false_fn(x, y):
                    return x - y

                return torch.cond(p, true_fn, false_fn, [a, b]) + a.sum()
```

Here are two versions of the output C++ codegen:

- ABI-compatible: https://gist.github.com/aakhundov/a8d14e4bc4bea7a4be388b5282afe3ea
- non-ABI-compatible: https://gist.github.com/aakhundov/88d6c7fc538bb8651e6e652ca11cb308

The full unit test is `test_aot_inductor.test_conditional_simple` in the current version of the PR.

File: torch/_inductor/codegen/cpp_wrapper_cpu.py
Comment (aakhundov) [inline torch/_inductor/codegen/cpp_wrapper_cpu.py:1452]: This is done (and the new ABI shim function `aoti_torch_copy_tensor` is added) to separate the ownership of the underlying `at::Tensor` instances between the inner and the outer scope: e.g., calling `.reset()` at the end of the codegened subgraph (inner scope) should decrement the refcount of the underlying storage, but not delete the storage, as it should still be accessible in the parent subgraph (outer scope).

**Review by jansel — APPROVED** [review]

File: torch/_inductor/codegen/cpp_wrapper_cpu.py
Comment (chenyang78) [inline torch/_inductor/codegen/cpp_wrapper_cpu.py:1481]: It seems that we might be able to leverage ```shared_ptr``` here, but I think it would require fair amount of work to pave the foundations.

**Review by chenyang78 — APPROVED** [review]
LGTM. Thanks!

**Comment by aakhundov** [comment]
The tests newly added in this PR are failing in ROCm periodic jobs. I think, it'd be better to disable these tests for now instead of reverting the PR. cc @izaitsevfb.

**Review by desertfire — COMMENTED** [review]
Forgot to hit "Submit review" previously. Left some minor comments.

File: torch/_inductor/codegen/cpp_wrapper_cpu.py
Comment (aakhundov) [reply torch/_inductor/codegen/cpp_wrapper_cpu.py:1352]: `codegen_int_array_var` relies on @lru_cache to cache on the arguments. The reason I've added the `graph` argument is to include the codegened graph into the cache key, hence make caching per codegened graph (or scope).

I don't think we can make the `self.get_codegened_graph()` the default value of the `graph` argument, as this will run into default being captured on the first pass (and not be changed in further calls). Therefore, I'm passing `self.get_codegened_graph()` from every call site to appropriately reflect the codegen graph at that call site in the arguments of `codegen_int_array_var`. Does this make sense, or do you mean something else?

File: torch/_inductor/codegen/cpp_wrapper_cpu.py
Comment (aakhundov) [reply torch/_inductor/codegen/cpp_wrapper_cpu.py:1475]: Not at this stage: the operands and outputs of the structured control flow operators (like `torch.cond` and `torch.while_loop`) are currently restricted to `torch.Tensor` type. If / when this is lifted, we can extend the codegen to deal with non-Tensor inputs / outputs.

File: torch/_inductor/codegen/wrapper.py
Comment (aakhundov) [reply torch/_inductor/codegen/wrapper.py:759]: Do you mean `EnterDeviceContextManagerLine` and `ExitDeviceContextManagerLine`? I think, they are doing `code.do_indent()` and `code.do_unindent()` directly (without relying on the `EnterSubgraphLine` and `ExitSubgraphLine`). Is that what you mean?

**Comment by jeffdaily** [comment]
> The tests newly added in this PR are failing in ROCm periodic jobs. I think, it'd be better to disable these tests for now instead of reverting the PR. cc @izaitsevfb.

@aakhundov how many rocm tests are you going to disable?  What's the threshold for newly failing tests being skipped instead of reverting the PR?  You've DISABLED 4 so far, but I need to DISABLE more.

**Comment by jeffdaily** [comment]
@aakhundov It is very disruptive to land a PR when you know there are CI failures.  Please don't do it again.  8 new DISABLED issues so far.

CC @malfet @huydhn @atalman

**Comment by aakhundov** [comment]
> It is very disruptive to land a PR when you know there are CI failures. Please don't do it again. 8 new DISABLED issues so far.

@jeffdaily  I apologize for the inconvenience.

The failures on ROCm weren't visible in this PR when I landed it: the `pytorchbot` comment above has indicated 1 "FLAKY" and 2 "BROKEN TRUNK" CI jobs (including the `ciflow/trunk` that I've added to the PR before merging it). So I haven't landed the PR knowing that there are failures.

I've discovered the ROCm failures only when I checked the HUD in the morning: my understanding is that they were coming from `ciflow/periodic`. Would you recommend to always add `ciflow/periodic` label to PRs before landing them?

Let me send a forward fix with `@skipIfRocm` on those tests so that we can close the DISABLED issues.

**Comment by aakhundov** [comment]
> how many rocm tests are you going to disable? What's the threshold for newly failing tests being skipped instead of reverting the PR? You've DISABLED 4 so far, but I need to DISABLE more.

As those tests are newly added in this PR, I think, we don't need to disable those tests if we skip them on ROCm? I'm going to send a forward fix adding `@skipIfRocm`.

**Comment by aakhundov** [comment]
@jeffdaily submitted a forward fix adding `@skipIfRocm` on all new tests in `test_aot_inductor.py`: https://github.com/pytorch/pytorch/pull/121522. I've added `ciflow/trunk` and `ciflow/periodic` labels on the PR. If we land this forward fix, I think, we can close the 8 new DISABLED issues?

**Comment by huydhn** [comment]
I think the correct response would be to just revert this PR and add `ciflow/rocm` to run rocm jobs before relanding it.  Creating multiple DISABLED issues is definitely an anti-pattern.  As the forward fix https://github.com/pytorch/pytorch/pull/121522 is already out, let's just quickly land it then.  Even then, there might also be more failing tests, so it could take few hours at least to confirm everything is green

**Comment by aakhundov** [comment]
@huydhn Thanks for the context! Next time will definitely revert.

**Comment by malfet** [comment]
> It is very disruptive to land a PR when you know there are CI failures. 

@jeffdaily I don't think this PR had any signal from ROCm at the merge time, so people are not clairvoyant so I think disable as either fwd-fix or thru the issue is an acceptable scenario there.
Are number of available ROCm runners(and queueing as result) is still the major blocker that prevents us from enabling testing as part of trunk job?

**Comment by malfet** [comment]
> I think the correct response would be to just revert this PR and add `ciflow/rocm` to run rocm jobs before relanding it.
Why? If something is not part of pull nor trunk, those failures would not block others and as such should be forward fixable, shouldn't it?

> Creating multiple DISABLED issues is definitely an anti-pattern. 

It's not great, but at the same time I'm not sure that just decorating more and more tests with `skipIfRocm` is a good solution either. As one can simply get into a habit of slapping it everywhere, out of abundance of caution.

**Comment by huydhn** [comment]
> Why? If something is not part of pull nor trunk, those failures would not block others and as such should be forward fixable, shouldn't it?

IMO, bringing `ciflow/rocm` back to be part of trunk is the goal.  May be we should consider bringing back the option to stamp `ciflow/rocm` when merging starts.

**Comment by aakhundov** [comment]
@jeffdaily @huydhn @malfet as there are more DISABLED tests coming, is it feasible to land the forward fix in https://github.com/pytorch/pytorch/pull/121522 now (without waiting for the CI)? I've tried `pytorchbot merge`, and it said there were 8 CI failures. Seems that those jobs have restarted and get cleared now. But there are still 70+ CI jobs to finish.

**Comment by aakhundov** [comment]
@jeffdaily I've manually closed the 11 DISABLED issues, as the forward fix in https://github.com/pytorch/pytorch/pull/121522 has landed.

**Comment by jeffdaily** [comment]
> > Why? If something is not part of pull nor trunk, those failures would not block others and as such should be forward fixable, shouldn't it?
> 
> IMO, bringing `ciflow/rocm` back to be part of trunk is the goal. May be we should consider bringing back the option to stamp `ciflow/rocm` when merging starts.

Getting full ROCm CI to run with each PR is the goal, but we aren't there yet with our capacity.  Our rough estimate is we would need somewhere between 2-4x our current capacity.  We have ~50 MI210 runners and ~45 MI50/60 runners.  The MI210 runners are exclusively in the linux.rocm.gpu.2 label and the MI50/60 runners are in the linux.rocm.gpu label, but that label also includes the same MI210 runners.  In other words, we have some tests we want run exclusively by MI210 and the rest of the tests could be either.

We have one short (~1.5hr) ROCm job that runs as part of all PRs.  We set this up as a concession instead of having the full ROCm CI suite run.  But it clearly isn't catching all failures because it doesn't cover everything.  When we submit a PR that exclusively touches ROCm features or issues, we add the ciflow/rocm label ourselves.  But if we have every PR add this label, including those that don't touch ROCm code paths, we would exhaust our capacity.  That's why we apply the label judiciously today.  Perhaps in the short term we should revisit the short-trunk-rocm job and add one of the test modules that is missing.  We're trying to balance getting a CI signal back in reasonable time versus coverage of important features.  Current list of trunk-rocm tests are

tests-to-include: test_nn test_torch test_cuda test_ops test_unary_ufuncs test_binary_ufuncs test_autograd inductor/test_torchinductor

**Comment by huydhn** [comment]
@jeffdaily Yeah, the capacity issue with ROCm is known to us, so we are also trying to approach the issue from another angle using target determination.  Basically, the CI needs to get smarter in deciding which tests to run and which to skip instead of having a fixed list of trunk-rocm tests.  Just FYI, we plan to roll out target determination for ROCm this half with the goal of cutting down the load there by 20%.

cc @clee2000 @osalpekar

**Pull request closed without merging**
